off-chain/src/Network/Types.purs

```diff
@@ -1,155 +1,207 @@
+newtype PendingAction a = PendingAction
+  { petitionID :: UserPetitionID
+  , action :: UserAction a
   }
 
-derive instance Newtype UserNodeState _
-derive instance Generic UserNodeState _
-
-instance Show UserNodeState where
-  show = genericShow
-
-instance Arbitrary UserNodeState where
-  arbitrary = do
-    numberOfRequests <- chooseInt 0 100000
-    pure $ UserNodeState { numberOfRequests }
-
--- | We split node data in three categories
--- | - Configuration, most functions won't perform changes on it.
--- | - State, we would keep changing it a lot in functions.
--- | - Information, read only things. 
+-- I suspect that this `a` is a type like :
+-- `UserPetitionID /\ UserAction a`
+newtype LeaderState a = LeaderState
+  { pendingActionsRequest :: AsyncMutableQueueRef a
+  -- For actions that were part of a previous built chain 
+  -- and were removed since a previous action failed.
+  , prioritaryPendingActions :: Array a
+  , signatureResponses :: AsyncMutableQueueRef a
+  , processedRequest :: MutableInt
+  , numberOfChainsBuilt :: ChainNumber
+  , stage :: LeaderServerStage
+  -- We really need to think if we really want this,
+  -- see `UserPetitionID` comment.
+  , numberOfActionsRequestsMade :: MutableInt
+  }
 
--- | The environment for user side
-newtype UserNode = UserNode
-  { configuration :: NodeConfiguration
-  , state :: UserNodeState
-  , information :: NodeInformation
+newtype LeaderConfiguration = LeaderConfiguration
+  { maxWaitingTimeForSignature :: MiliSeconds
+  , maxQueueSize :: Int
+  , numberOfActionToTriggerChainBuilder :: Int
+  , maxWaitingTimeBeforeBuildChain :: Int
+  -- Those are built on top of a network or IO interface
+  , clientHandlers ::
+      { transactionSignature :: AskForSignature -> Aff $ Either String Unit
+      , reportError :: ReportErrorAtProcessing -> Aff $ Either String Unit
+      }
   }
 
-derive instance Newtype UserNode _
-derive instance Generic UserNode _
+newtype LeaderNode a = LeaderNode
+  { state :: LeaderState a
+  , configuration :: LeaderConfiguration
+  }
 
--- | The environment for leader side
-newtype LeaderNode = LeaderNode
-  { configuration :: NodeConfiguration
-  , state :: LeaderNodeState
-  , information :: NodeInformation
+newtype UserState a = UserState
+  {
+    -- | This three types must store the state of the sent transactions.
+    -- | `actionSent` is for actions that are confirmed to be received
+    -- | by the leader but aren't in processes right now.
+    -- TODO : put the right types in the following three
+    -- Maybe is : `MutableReference (OrderedMap (SomeUniqueID a) UserAction a)
+    actionsSent :: Array a
+  -- | `transactionsSent` is intended for both the `action` and it's
+  -- | corresponding `SignedTransaction` already send to the server.
+  , transactionsSent :: Array a
+  -- | Those signed transactions are confirmed to be in the chain 
+  -- | ready for submission.
+  -- | Once a transaction is confirmed to be done, we can safely 
```

Comment:
> Once a transaction is confirmed to be done

What "done" means here? Submitted to the node by the leader (via whatever submits it to the node in CLT) ?